src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProjectFactory.cs

```diff
@@ -217,9 +221,25 @@ public async ValueTask ApplyChangeToWorkspaceMaybeAsync(bool useAsync, Action<Wo
         }
     }
 
+    /// <summary>
+    /// Applies a single operation to the workspace that also needs to update the <see cref="_projectUpdateState"/>.
+    /// <paramref name="action"/> should be a call to one of the protected Workspace.On* methods.
+    /// </summary>
+    public void ApplyChangeToWorkspaceWithProjectUpdateState(Func<Workspace, ProjectUpdateState, ProjectUpdateState> action)
+    {
+        using (_gate.DisposableWait())
+        {
+            var projectUpdateState = _projectUpdateState;
+            projectUpdateState = action(Workspace, projectUpdateState);
+            ApplyProjectUpdateState(projectUpdateState);
```

Comment:
nit.  have different locals.  So like: `var initialProjectUpdateState;` and `var finalProjectUpdateState`